hebiros_gazebo_plugin/sim/joint.cpp

```diff
@@ -0,0 +1,347 @@
+      intermediate_effort = effort_cmd_ + position_pid_out + velocity_pid_out;
+      pwm_cmd_ = Clip(
+        effort_pid_.update(intermediate_effort, effort_fbk_, dt),
+        MIN_PWM, MAX_PWM);
+      break;
+
+    case ControlStrategy::Strategy3:
+      position_pid_out = Clip(
+        position_pid_.update(position_cmd_, position_fbk_, dt),
+        MIN_PWM, MAX_PWM);
+      velocity_pid_out = Clip(
+        velocity_pid_.update(velocity_cmd_, velocity_fbk_, dt),
+        MIN_PWM, MAX_PWM);
+      effort_pid_out = Clip(
+        effort_pid_.update(effort_cmd_, effort_fbk_, dt),
+        MIN_PWM, MAX_PWM);
+      pwm_cmd_ = Clip(position_pid_out + velocity_pid_out + effort_pid_out, MIN_PWM, MAX_PWM);
+      break;
+
+    case ControlStrategy::Strategy4:
+      position_pid_out = position_pid_.update(position_cmd_, position_fbk_, dt);
+      intermediate_effort = effort_cmd_ + position_pid_out;
+      effort_pid_out = Clip(
+        effort_pid_.update(intermediate_effort, effort_fbk_, dt),
+        MIN_PWM, MAX_PWM);
+      velocity_pid_out = Clip(
+        velocity_pid_.update(velocity_cmd_, velocity_fbk_, dt),
+        MIN_PWM, MAX_PWM);
+      pwm_cmd_ = Clip(velocity_pid_out + effort_pid_out, MIN_PWM, MAX_PWM);
+      break;
+
+    default:
+      pwm_cmd_ = 0;
+  }
+ 
+  // Apply safety limits
+  pwm_cmd_ = temperature_safety_.limit(pwm_cmd_);
+}
+  
+double Joint::generateForce(double dt) {
+  //Set target positions
+  double force, alpha;
+
+  float voltage = 48.0f;
+  float motor_velocity = velocity_fbk_ * gear_ratio_;
+  float speed_constant = 1530.0f;
+  float term_resist = 9.99f;
+  if (isX8(joint_type_)) {
+    speed_constant = 1360.0f;
+    term_resist = 3.19f;
+  }
+
+  if (pwm_cmd_ == 0) {
+    force = 0;
+  }
+  else {
+    // TODO: use temp compensation here, too?
+    force = ((pwm_cmd_ * voltage - (motor_velocity / speed_constant)) / term_resist) * 0.00626 * gear_ratio_ * 0.65;
+  }
+
+  float prev_winding_temp = temperature_.getMotorWindingTemperature();
+
+  // Get components of power into the motor
+
+  // Temperature compensated speed constant
+  float comp_speed_constant = speed_constant * 1.05f * // Experimental tuning factor                           
+    (1.f + .001f * (prev_winding_temp - 20.f)); // .001 is speed constant change per temperature change 
+  float winding_resistance = term_resist * 
+    (1.f + .004f * (prev_winding_temp - 20.f)); // .004 is resistance change per temperature change for copper 
+  float back_emf = (motor_velocity * 30.f / M_PI) / comp_speed_constant;
+  float winding_voltage = pwm_cmd_ * voltage - back_emf;
+
+  // TODO: could add ripple current estimate here, too
+
+  // Update temperature:
+  // Power = I^2R, but I = V/R so I^2R = V^2/R:
+
+  double power_in = winding_voltage * winding_voltage / winding_resistance;
+  temperature_.update(power_in, dt);
+  temperature_safety_.update(temperature_.getMotorWindingTemperature());
+
+  // Low pass?
+
+  //alpha = 0.1;
```

Comment:
Is this intentionally commented out?

Reply:
Yeah -- this is some of the code that can optionally be commented out to improve effort control.  Still an active area of work in progress, but will be easier to work on once the code is broken out so we can set up a better test bed.